gnomad_qc/v4/analyses/grpmax_comps.py

```diff
@@ -0,0 +1,292 @@
+
+logging.basicConfig(format="%(levelname)s (%(name)s %(lineno)s): %(message)s")
+logger = logging.getLogger("grpmax_comps")
+logger.setLevel(logging.INFO)
+
+
+DIVERSE_GRPS = hl.literal({"afr", "amr", "eas", "mid", "sas"})
+EUR_GRPS = {"all_eur": ["nfe", "fin", "asj"], "nfe_only": ["nfe"]}
+FILTER_VALUES_TO_DROP = hl.array(["AC0", "InbreedingCoeff"])
+AF_THRESHOLDS = [0.0001, 0.001]
+NS_CONSEQ_TERMS = hl.literal(CSQ_CODING_HIGH_IMPACT + CSQ_CODING_MEDIUM_IMPACT)
+
+
+def get_eur_freq(ht: hl.Table, eur_grps: list, version: str = "v4"):
+    """
+    Calculate European genetic ancestry super group AF.
+
+    :param ht: gnomAD release Table
+    :param eur_grps: Set of genetic ancestry groups to be included in AF
+    :param version: gnomAD version
+    :return: Table with European AF annotation
+    """
+    if version == "v4":
+        eur_indices = hl.literal(eur_grps).map(lambda g: ht.freq_index_dict[g + "_adj"])
+    else:
+        eur_indices = hl.literal(eur_grps).map(
+            lambda g: ht.freq_index_dict["gnomad_" + g]
+        )
+    ht = ht.annotate(
+        eur_AC=hl.sum(
+            eur_indices.map(
+                lambda f: hl.if_else(hl.is_defined(ht.freq[f].AC), ht.freq[f].AC, 0)
+            )
+        ),
+        eur_AN=hl.sum(
+            eur_indices.map(
+                lambda f: hl.if_else(hl.is_defined(ht.freq[f].AN), ht.freq[f].AN, 0)
+            )
+        ),
+    )
+    ht = ht.annotate(eur_AF=hl.if_else(ht.eur_AN > 0, ht.eur_AC / ht.eur_AN, 0.0))
+    return ht.drop("eur_AC", "eur_AN")
+
+
+def filter_to_threshold(
+    ht: hl.Table,
+    af_threshold: float = 0.001,
+    version: str = "v4",
+):
+    """
+    Filter to variants where eur AF is < threshold while grpmax > threshold.
+
+    :param ht: gnomAD release Table
+    :param af_threshold: AF dividing threshold
+    :param version: gnomAD version
+    :return: Table filtered to variants meething threshold specifications
+    """
+    if version == "v4":
+        grpmax_expr = ht.grpmax.gnomad
+    else:
+        grpmax_expr = ht.popmax[0]
+    ht = ht.filter((grpmax_expr.AF > af_threshold) & (ht.eur_AF < af_threshold))
+    return ht
+
+
+def version_stats(
+    ht: hl.Table, version: str = "v4", can_only: bool = False
+) -> Dict[str, Dict[str, Dict[str, int]]]:
+    """
+    Calculate grpmax stats for a given gnomAD version.
+
+    :param ht: gnomAD release Table
+    :param version: gnomAD version
+    :param can_only: Only consider MANE Select and canonical transcripts.
+    :return: Dictionary of grpmax stats
+    """
+    logger.info("Calculating grpmax stats for %s", version)
+    # Group results into nfe and all eur grps
+    results_by_eur_grping = {}
+
+    logger.info(
+        "Total number of variants in %s before filtering: %s", version, ht.count()
+    )
+    t_variants = ht.count()
+    if can_only:
+        logger.info("Filtering to only MANE Select and canonical transcripts")
+        ht = ht.explode(ht.vep.transcript_consequences)
```

Comment:
you should be able to filter to canonical transcripts using https://github.com/broadinstitute/gnomad_methods/blob/main/gnomad/utils/vep.py#L393 (something like `filter_vep_to_canonical_transcripts(ht, filter_empty_csq=True)`) without having to explode here (which would also remove the `distinct` downstream)